### PR TITLE
Channel Establishment for V3 Channels

### PR DESCRIPTION
**Pull request by carlaKC**
This PR updates the channel establishment flow to allow and validate V3 channels (behind `test` flag).

### CODE REVIEW

**Review by tankyleo — COMMENTED** [review]
Sorry for the delay, here's a first pass :)

File: lightning/src/ln/channelmanager.rs [hunk 1]
```diff
@@ -16123,6 +16123,13 @@ mod tests {
 		assert_eq!(accept_message.channel_reserve_satoshis, 2_000);
 	}
 
+	#[test]
+	fn test_inbound_zero_fee_commitments_acceptance() {
```
Comment (tankyleo) [inline lightning/src/ln/channelmanager.rs:16127]: nit: we have `test_inbound_anchors_manual_acceptance` above, so should this be `test_inbound_zero_fee_commitments_manual_acceptance` ? For consistency ?

File: lightning/src/ln/channelmanager.rs [hunk 2]
```diff
@@ -16123,6 +16123,13 @@ mod tests {
...
+		let mut zero_fee_cfg = test_default_channel_config();
+		zero_fee_cfg.channel_handshake_config.negotiate_anchor_zero_fee_commitments = true;
+		do_test_manual_inbound_accept_with_override(zero_fee_cfg, None);
```
Comment (tankyleo) [inline lightning/src/ln/channelmanager.rs:16130]: Should we also test `zero_fee_cfg` *with* overrides ?

**Review by tankyleo — COMMENTED** [review]
Another pass :)

File: lightning/src/ln/channelmanager.rs (hunk 2 above)
Comment (carlaKC) [reply lightning/src/ln/channelmanager.rs:16130]: Doesn't really offer us much more coverage since [config override](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/ln/channelmanager.rs#L7878) is independent of the channel type?

File: lightning/src/ln/channel.rs [hunk 3]
```diff
@@ -4860,7 +4860,16 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
...
-		if channel_type.supports_anchors_zero_fee_htlc_tx() {
+		if channel_type.supports_anchor_zero_fee_commitments() {
+			channel_type.clear_anchor_zero_fee_commitments();
+			channel_type.set_anchors_zero_fee_htlc_tx_required();
+			channel_type.set_static_remote_key_required();
+
+			self.feerate_per_kw = fee_estimator.bounded_sat_per_1000_weight(ConfirmationTarget::AnchorChannelFee);
+			assert!(!channel_type.supports_anchor_zero_fee_commitments());
+			assert!(channel_type.supports_anchors_zero_fee_htlc_tx());
+			assert!(channel_type.supports_static_remote_key());
```
[the remark this answers is not in the document]
Comment (carlaKC) [reply lightning/src/ln/channel.rs:4877]: Oh yeah, good catch! I think that this is actually an issue on both ends?
- We have to check that we want to negotiate this type
- We need to check that the remote _actually_ supports `anchors_zero_fee_htlc_tx_required`

Previously this wasn't an issue because the `static_remote` is a dependency of `anchors_zero_fee_htlc_tx` - if they support `anchors_zero_fee_htlc_tx` we know that we can drop down to `static_remote`. 

But now if they've set `anchor_zero_fee_commitments` we can't just assume that they support `anchors_zero_fee_htlc_tx` without going back to look at `InitFeatures`?

File: lightning/src/ln/channel.rs [hunk 4]
```diff
@@ -3036,12 +3036,18 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
...
+				(0, 0)
+			} else if channel_type.supports_anchors_zero_fee_htlc_tx() {
+				let feerate = fee_estimator
+					.bounded_sat_per_1000_weight(ConfirmationTarget::AnchorChannelFee);
+				(feerate, ANCHOR_OUTPUT_VALUE_SATOSHI * 2 * 1000)
+			} else {
+				let feerate = fee_estimator
+					.bounded_sat_per_1000_weight(ConfirmationTarget::NonAnchorChannelFee);
+				(feerate, 0)
+			};
```
[the remark this answers is not in the document]
Comment (tankyleo) [reply lightning/src/ln/channel.rs:3050]: Ah yes ! All good, feel free to add a comment; it was sloppy on my side, staring a the code again made it click :)

I know we asserted that `open_channel_fields.commitment_feerate_sat_per_1000_weight` was 0 for zero-fee commitments further above via `check_remote_fee` , but wondering if we should add a quick debug assert right before the "can't even pay for initial commitment transaction fee of x sats" in `new_for_inbound_channel` ? Something like this?
```diff
+ // Sanity check that `commitment_tx_fee` is 0 for zero fee commitments
+ debug_assert!(!channel_type.supports_anchor_zero_fee_commitments()
+     || commitment_tx_fee == 0)
if (funders_amount_msat / 1000).saturating_sub(anchor_outputs_value) < commitment_tx_fee {
```

File: lightning/src/ln/channel.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tankyleo) [reply lightning/src/ln/channel.rs:4877]: > I think that this is actually an issue on both ends?

Yes !

> But now if they've set anchor_zero_fee_commitments we can't just assume that they support anchors_zero_fee_htlc_tx without going back to look at InitFeatures?

Could we assume they have set it, and if not, we get another error back, and we come back here to downgrade to static remote key ?

File: lightning/src/ln/channelmanager.rs (hunk 2 above)
Comment (tankyleo) [reply lightning/src/ln/channelmanager.rs:16130]: Ah yes I see it now thank you.

IIUC then this test right here double checks that if `mannual_accept_cfg.manually_accept_inbound_channels = false`, and `channel_handshake_config.negotiate_anchor_zero_fee_commitments = true`, that the receiver node complains with "No channels with anchor outputs accepted.". Also checks that when both of these settings are true, we *do* accept the channel.

File: lightning/src/ln/channel.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (carlaKC) [reply lightning/src/ln/channel.rs:3050]: > debug_assert!(!channel_type.supports_anchor_zero_fee_commitments() || commitment_tx_fee == 0)

I don't think that this adds much, given the checks that we have in `check_remote_fee`?
- We've already errored out if `supports_anchor_zero_fee_commitments` isn't zero fee
- All other channel types are checked against `bounded_sat_per_1000_weight` (which can't be smaller than `FEERATE_FLOOR_SATS_PER_KW=253`), which implicitly checks that they're non-zero

File: lightning/src/ln/channel.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (carlaKC) [reply lightning/src/ln/channel.rs:4877]: > Could we assume they have set it, and if not, we get another error back, and we come back here to downgrade to static remote key ?

I don't _love_ spending a round trip on a channel type that we know is going to fail. Looking at re-using the code in `get_initial_channel_type` for downgrades to DRY this up a bit. Downside is that we need to thread remote features through, but I think it's a nicer approach - WDYT?

[the remark this answers is not in the document]
Comment (tankyleo) [reply lightning/src/ln/channel.rs:4877]: Yes I don't like that roundtrip either :)

File: lightning/src/ln/channel.rs [hunk 5]
```diff
@@ -4854,21 +4866,45 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
...
+			assert!(!eligible_features.supports_anchor_zero_fee_commitments());
+		}
+
+		let next_channel_type = get_initial_channel_type(user_config, &eligible_features);
+
+		// Note that we can't get `anchor_zero_fee_commitments` type here, which requires zero
+		// fees, because we downgrade from this channel type first. If there were a superior
+		// channel type that downgrades to `anchor_zero_fee_commitments`, we'd need to handle
+		// fee setting differently here.
+		assert!(!next_channel_type.supports_anchor_zero_fee_commitments());
```
Comment (tankyleo) [inline lightning/src/ln/channel.rs:4899]: A note: I think Matt mentioned in the past to avoid hard asserts unless we are at risk of losing funds (and instead favor debug asserts). This assert makes sense to me though :)

File: lightning/src/ln/channelmanager.rs [hunk 6]
```diff
@@ -16153,41 +16170,122 @@ mod tests {
...
+		do_test_channel_type_downgrade(initiator_cfg, receiver_cfg, start_type, downgrade_types);
+	}
+
+	#[test]
+	fn test_zero_fee_commitments_downgrade() {
+		// Tests that the local node will retry without zero fee commitments in the case where the
+		// remote node supports the feature but does not accept it.
+		let mut initiator_cfg = test_default_channel_config();
+		initiator_cfg .channel_handshake_config.negotiate_anchor_zero_fee_commitments = true;
+		initiator_cfg .channel_handshake_config.negotiate_anchors_zero_fee_htlc_tx = true;
```
Comment (tankyleo) [inline lightning/src/ln/channelmanager.rs:16224]: nit: Were these spaces intentional ?

File: lightning/src/ln/channelmanager.rs [hunk 7]
```diff
@@ -16153,41 +16170,122 @@ mod tests {
...
+	fn test_zero_fee_commitments_downgrade_to_static_remote() {
+		// Tests that the local node will retry with static remote key when zero fee commitments
+		// are supported (but not accepted), but not legacy anchors.
+		let mut initiator_cfg = test_default_channel_config();
+		initiator_cfg .channel_handshake_config.negotiate_anchor_zero_fee_commitments = true;
+		initiator_cfg .channel_handshake_config.negotiate_anchors_zero_fee_htlc_tx = true;
+
+		let mut receiver_cfg = test_default_channel_config();
+		receiver_cfg.channel_handshake_config.negotiate_anchor_zero_fee_commitments = true;
+		receiver_cfg.manually_accept_inbound_channels = true;
```
Comment (tankyleo) [inline lightning/src/ln/channelmanager.rs:16249]: Do we also want to run a version of this test where *local* does *not* support zero fee htlc (instead of *remote*) ?

File: lightning/src/ln/channelmanager.rs (hunk 6 above)
Comment (carlaKC) [reply lightning/src/ln/channelmanager.rs:16224]: Nope! 

(I am weak and accustomed to fmt on save 😅 )

File: lightning/src/ln/channel.rs (hunk 5 above)
Comment (carlaKC) [reply lightning/src/ln/channel.rs:4899]: Good to know! Not familiar w/ the conventions for the project.

Will leave this as-is pending input from a second reviewer.

File: lightning/src/ln/channelmanager.rs (hunk 7 above)
Comment (tankyleo) [reply lightning/src/ln/channelmanager.rs]: I might have been unclear, I meant a diff similar to this one ? Or do you think it's not necessary ?

Thank you thank you :) 

```diff
                let start_type = ChannelTypeFeatures::anchors_zero_fee_commitments();
                let end_type = ChannelTypeFeatures::only_static_remote_key();
                do_test_channel_type_downgrade(initiator_cfg, receiver_cfg, start_type, vec![end_type]);
+
+               let mut initiator_cfg = test_default_channel_config();
+               initiator_cfg.channel_handshake_config.negotiate_anchor_zero_fee_commitments = true;
+
+               let mut receiver_cfg = test_default_channel_config();
+               receiver_cfg.channel_handshake_config.negotiate_anchor_zero_fee_commitments = true;
+               receiver_cfg.channel_handshake_config.negotiate_anchors_zero_fee_htlc_tx = true;
+               receiver_cfg.manually_accept_inbound_channels = true;
+
+               let start_type = ChannelTypeFeatures::anchors_zero_fee_commitments();
+               let end_type = ChannelTypeFeatures::only_static_remote_key();
+               do_test_channel_type_downgrade(initiator_cfg, receiver_cfg, start_type, vec![end_type]);
        }
```

Comment (carlaKC) [reply lightning/src/ln/channelmanager.rs]: I think we're okay without this test because we already have coverage for a downgrade from `zero_fee_commitments` to `static_remote_key` in `test_zero_fee_commitments_downgrade_to_static_remote`. 

My goal with these tests was to cover downgrade logic, rather than indirectly test the different ways we can get channel types out of `get_initial_channel_type`.

Comment (tankyleo) [reply lightning/src/ln/channelmanager.rs]: Discussed offline - the diff above really belongs in tests for `get_initial_channel_type`

File: lightning/src/ln/channel.rs (hunk 5 above)
Comment (TheBlueMatt) [reply lightning/src/ln/channel.rs:4932]: Indeed, if we think we can lose funds we should hard-assert, if we don't, we should debug-assert. If we get 0FC here we'd presumably set a fee, and then I assume that implies we'd be vulnerable to funds loss (as the commitment would be tagged TRUC but not 0F and might have a zero-value anchor, making it ineligible for relay under policy)? If that's the theory, makes sense, but might be worth highlighting the risks a bit clearer in the comment.

Comment (carlaKC) [reply lightning/src/ln/channel.rs:4932]: Indeed, if we got `next_channel_type.supports_anchor_zero_fee_commitments` here, we'd try to open a 0F channel with non-zero fees. 

We'd probably just fail to open the channel because our peer would reject it, but probably don't want to rely on that (something, something, attacker lets us create this unrelayable commitment and then uses an accelerator?).

Will leave `assert` and update the comment 👍

**Comment by carlaKC** [comment]
Rebased + added additional comment on assert - [diff](https://github.com/lightningdevkit/rust-lightning/compare/e56e32ab04fdaab110d132853d4dab1f0f028802..5b568c38def52c3ab4a6fe005fb96228f384381b).

**Comment by carlaKC** [comment]
Two followups for this to do some test cleaning while we're here:
- #3827: moves + formats some of the `channel.rs` tests and adds coverage for `get_initial_channel_type` as disucssed with @tankyleo
- #3797: moves + formats some channel downgrades from `channelmanager.rs`

File: lightning/src/ln/channel.rs [hunk 8]
```diff
@@ -4854,21 +4866,48 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
...
+			eligible_features.clear_anchors_zero_fee_htlc_tx();
+			assert!(!eligible_features.supports_anchor_zero_fee_commitments());
+			assert!(!eligible_features.supports_anchors_nonzero_fee_htlc_tx());
 		} else if channel_type.supports_scid_privacy() {
-			channel_type.clear_scid_privacy();
+			eligible_features.clear_scid_privacy();
+			eligible_features.clear_anchors_zero_fee_htlc_tx();
+			eligible_features.clear_anchor_zero_fee_commitments();
+			assert!(!eligible_features.supports_scid_privacy());
+			assert!(!eligible_features.supports_anchors_nonzero_fee_htlc_tx());
```
Comment (tankyleo) [inline lightning/src/ln/channel.rs]: should have raised this nit earlier sorry: for consistency shouldn't we *also* add this here and in the previous branch ?
```
assert!(!eligible_features.supports_anchors_zero_fee_htlc_tx());
```

Also before this commit, we were just sanity checking that we aren't supporting the deprecated `nonzero_fee_htlc_tx` type, but now we also check that the earlier clear statements worked correctly ?

Comment (carlaKC) [reply lightning/src/ln/channel.rs]: > assert!(!eligible_features.supports_anchors_zero_fee_htlc_tx());

Nice catch!

> Also before this commit, we were just sanity checking that we aren't supporting the deprecated nonzero_fee_htlc_tx type, but now we also check that the earlier clear statements worked correctly ?

I was worried about not clearing out the feature bits in the correct order and ending up in a weird loop where we downgrade to one channel type, don't clear the feature bits and then "downgrade" back to a type we already tried - which would mean we endlessly retry the channel. So we want to be really sure that we're clearing all of the features for our channel type and "greater".

Perhaps it's overkill/should be debug?

**Comment by carlaKC** [comment]
[Rebased](https://github.com/lightningdevkit/rust-lightning/compare/5b568c38def52c3ab4a6fe005fb96228f384381b..e4302126a6b38eb19949f765bbe56b10ac272714) + [addressed nit](https://github.com/lightningdevkit/rust-lightning/compare/e4302126a6b38eb19949f765bbe56b10ac272714..99c645e4b37411c2ce914ac161392f57db881f1a).

File: lightning/src/ln/channel.rs (hunk 8 above)
Comment (tankyleo) [reply lightning/src/ln/channel.rs]: > I was worried about not clearing out the feature bits in the correct order and ending up in a weird loop where we downgrade to one channel type, don't clear the feature bits and then "downgrade" back to a type we already tried - which would mean we endlessly retry the channel. So we want to be really sure that we're clearing all of the features for our channel type and "greater".

All sounds good to me thank you carla - we can leave those asserts as they are.

**Review by tankyleo — APPROVED** [review]

File: lightning/src/ln/channel.rs [hunk 9]
```diff
@@ -4889,21 +4901,50 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
...
+		let mut eligible_features = their_features.clone();
+
+		if channel_type.supports_anchor_zero_fee_commitments() {
+			eligible_features.clear_anchor_zero_fee_commitments();
+			assert!(!eligible_features.supports_anchor_zero_fee_commitments());
+		} else if channel_type.supports_anchors_zero_fee_htlc_tx() {
+			eligible_features.clear_anchor_zero_fee_commitments();
+			eligible_features.clear_anchors_zero_fee_htlc_tx();
+			assert!(!eligible_features.supports_anchor_zero_fee_commitments());
+			assert!(!eligible_features.supports_anchors_nonzero_fee_htlc_tx());
```
Comment (TheBlueMatt) [inline lightning/src/ln/channel.rs]: Isn't this trivially reachable if a peer sets the nonzero-fee-htlc-tx feature bit? Generally, though, all the assertions here seem like they could be debug_assertions or elided entirely (asserting that we don't support something we just cleared probably belongs in a features test, and I believe already exists).

Comment (carlaKC) [reply lightning/src/ln/channel.rs]: Will remove the feature bit asserts 👍  

> Isn't this trivially reachable if a peer sets the nonzero-fee-htlc-tx feature bit?

ofc, it should be on `channel_type` 🤦‍♀️

File: lightning/src/ln/channel.rs [hunk 10]
```diff
@@ -4885,25 +4891,35 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
 			// We've exhausted our options
 			return Err(());
 		}
+
+		// We should never have negotiated `anchors_nonzero_fee_htlc_tx` because it can result in a
+		// loss of funds.
+		let channel_type = &funding.channel_transaction_parameters.channel_type_features;
+		assert!(!channel_type.supports_anchors_nonzero_fee_htlc_tx());
```
Comment (carlaKC) [inline lightning/src/ln/channel.rs:4904]: Moved this assert up because we should never negotiate a non-zero-htlc-anchor channel, and it's a risk of lost funds if we do.

Couldn't see a historical reason why we only checked it if the channel type was `supports_anchors_zero_fee_htlc_tx` - went back a few commits and couldn't find reasoning. cc @TheBlueMatt

**Comment by carlaKC** [comment]
Removed asserts + moved `supports_anchors_nonzero_fee_htlc_tx` assert, [diff here](https://github.com/lightningdevkit/rust-lightning/compare/99c645e4b37411c2ce914ac161392f57db881f1a..23de01cd59a4899b327c0b6f99a0a542a97d0232).

File: lightning/src/ln/channel.rs [hunk 11]
```diff
@@ -4885,25 +4897,47 @@ impl<SP: Deref> ChannelContext<SP> where SP::Target: SignerProvider {
...
+		assert!(!next_channel_type.supports_anchor_zero_fee_commitments());
+		let conf_target = if next_channel_type.supports_anchors_zero_fee_htlc_tx() {
+			ConfirmationTarget::AnchorChannelFee
 		} else {
-			*channel_type = ChannelTypeFeatures::only_static_remote_key();
-		}
+			ConfirmationTarget::NonAnchorChannelFee
+		};
+		self.feerate_per_kw = fee_estimator.bounded_sat_per_1000_weight(conf_target);
+	 	funding.channel_transaction_parameters.channel_type_features = next_channel_type;
```
Comment (tankyleo) [inline lightning/src/ln/channel.rs:4939]: Sorry do we want to fix this ? 

```
lightning/src/ln/channel.rs:4939: space before tab in indent.
+               funding.channel_transaction_parameters.channel_type_features = next_channel_type;
```
